File: probs/826.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        vector<pair<int, int> > jobs;
        for (int i = 0; i < difficulty.size(); i++) {
            jobs.push_back({ difficulty[i], profit[i] });
        }

        sort(jobs.begin(), jobs.end());
        sort(worker.begin(), worker.end());

        int job = 1;

        pair<int, int> best = jobs[0];
        int result = 0;

        for (int i = 0; i < worker.size(); i++) {
            while (job < jobs.size() && jobs[job].first <= worker[i]) {
                if (jobs[job].second > best.second) {
                    best = jobs[job];
                }
                
                job++;
            }

            if (best.first <= worker[i]) {
                result += best.second;
            }
        }

        return result;
    }
};

File: probs/826.rs
impl Solution {
    pub fn max_profit_assignment(difficulty: Vec<i32>, profit: Vec<i32>, mut worker: Vec<i32>) -> i32 {
        let mut jobs: Vec<(i32, i32)> = difficulty.iter()
            .zip(profit.iter())
            .map(|(&d, &p)| (d, p))
            .collect();

        jobs.sort();
        worker.sort();

        let mut best: (i32, i32) = jobs[0];
        let mut job = 0;
        let mut result = 0;

        for w in worker.iter() {
            while job < jobs.len() && jobs[job].0 <= *w {
                if jobs[job].1 > best.1 {
                    best = jobs[job];
                }

                job += 1;
            }

            if best.0 <= *w {
                result += best.1;
            } 
        }

        result
    }
}
